### PR TITLE
Re-running RAVEN workflows in Python/Jupyter environments

### PR DESCRIPTION
**Pull request by dgarrett622**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
#1682 

##### What are the significant changes in functionality due to this change request?
Enable re-running RAVEN workflows in Python or Jupyter notebook environments

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [ ] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by dgarrett622** [comment]
Resolving conflicts in the conflicting files does not solve the issue. I'll need to spend more time on this.

File: ravenframework/Databases/Database.py [hunk 1]
```diff
@@ -121,22 +119,21 @@ def _handleInput(self, paramInput):
...
 
-  def initialize(*args, **kwargs):
+  def initialize(self, *args, **kwargs):
     """
       Initialization for data object, if any.
       @ In, args, list, ordered arguments
       @ In, kwargs, dict, keyword arguments
       @ Out, None
     """
-    pass
```
Comment (PaulTalbot-INL) [inline ravenframework/Databases/Database.py:139]: I like the "pass" for explicit clarity, but to preference.

File: ravenframework/MessageHandler.py [hunk 2]
```diff
@@ -19,19 +19,11 @@
...
 import time
 import bisect
-import builtins
 
 from .utils import utils
 
 _starttime = time.time()
 
-#custom exceptions
-class NoMoreSamplesNeeded(GeneratorExit):
```
Comment (PaulTalbot-INL) [inline ravenframework/MessageHandler.py:29]: watching for where this got moved to ...

File: ravenframework/MessageHandler.py [hunk 3]
```diff
@@ -66,10 +58,6 @@ class NoMoreSamplesNeeded(GeneratorExit):
 
 TL;DR: BaseClasses/MessageUser is a superclass that gives access to hooks to the simulation's MessageHandler
 instance, while the MessageHandler is an output stream control tool.
-
-In an effort to make the MH more flexible, we insert getMessageHandler into the python "builtins" module.
-This means that any time after this module (MessageHandler) is imported, you can use
-"getMessageHandler(name='default')" to retrieve a particular message handler as identified by "name".
```
Comment (PaulTalbot-INL) [inline ravenframework/MessageHandler.py:72]: this was an intentional change requested by a previous PR; is there a critical reason to change it?

File: ravenframework/MessageHandler.py [hunk 4]
```diff
@@ -92,28 +80,33 @@ def __init__(self):
...
       Initializes basic instance attributes
       @ In, initDict, dict, dictionary of global options
       @ Out, None
     """
-    self.verbosity = initDict.get('verbosity','all').lower()
-    self.callerLength = initDict.get('callerLength',25)
-    self.tagLength = initDict.get('tagLength',15)
+    self.verbosity    = initDict.get('verbosity', 'all').lower()
+    self.callerLength = initDict.get('callerLength', 25)
+    self.tagLength    = initDict.get('tagLength', 15)
```
Comment (PaulTalbot-INL) [inline ravenframework/MessageHandler.py:109]: aligning "=" isn't expressly forbidden, but it is non-pep8. In generally we've been removing sign alignments and favoring comment alignments.

File: ravenframework/Models/HybridModels/HybridModel.py [hunk 5]
```diff
@@ -178,6 +179,11 @@ def initialize(self,runInfo,inputs,initDict=None):
       if romIn.amITrained:
         self.raiseAWarning("The provided rom ", romIn.name, " is already trained, we will reset it!")
         romIn.reset()
+        # reset HybridModel attributes to allow workflow re-running
+        self.romConverged = False
+        self.oldTrainingSize = 0
+        self.modelIndicator = {}
+        self.crowdingDistance = None
```
Comment (PaulTalbot-INL) [inline ravenframework/Models/HybridModels/HybridModel.py:186]: could we collect this in a particular reset method so it's obvious where to expand these if we need to in the future?

File: ravenframework/Optimizers/GradientDescent.py [hunk 6]
```diff
@@ -126,9 +126,9 @@ def getInputSpecification(cls):
               it can be accepted. Exactly one of the acceptance criteria
               below may be selected for this Optimizer.""")
     specs.addSub(accept)
-    ## common options to all acceptanceCondition descenders
-    ## TODO
-    ## get specs for each acceptanceCondition subclass, and add them to this class's options
+    # common options to all acceptanceCondition descenders
+    # TODO
+    # get specs for each acceptanceCondition subclass, and add them to this class's options
```
Comment (PaulTalbot-INL) [inline ravenframework/Optimizers/GradientDescent.py:131]: I think we can remove this TODO, as the next couple lines appear to do just that.

File: ravenframework/Optimizers/GradientDescent.py [hunk 7]
```diff
@@ -689,15 +717,16 @@ def _addToSolutionExport(self, traj, rlz, acceptable):
     """
     toAdd = {'stepSize': self._stepHistory[traj][-1]['magnitude']}
     for key, val in self._convergenceInfo[traj].items():
-      toAdd['conv_{}'.format(key)] = val
+      toAdd[f'conv_{key}'] = val
     # collect any additions from gradient and stepper
-    ## gradient
```
Comment (PaulTalbot-INL) [inline ravenframework/Optimizers/GradientDescent.py:694]: I can see you have strong preferences on double comments; I use them as a heirarchy so I can see which comments are sub-bullets of other comments. I recommend trying to keep a light touch on other dev's style choices when going over code.

File: ravenframework/OutStreams/PlotInterfaces/GeneralPlot.py [hunk 8]
```diff
@@ -136,204 +95,157 @@ def __init__(self):
...
     self.clusterValues = None
 
-    ## Gaussian Mixtures
+    # Gaussian Mixtures
     self.mixtureLabels = None
     self.mixtureValues = None
     self.mixtureMeans = None
     self.mixtureCovars = None
 
-  # TODO started, but didn't finish due to time constraints
```
Comment (PaulTalbot-INL) [inline ravenframework/OutStreams/PlotInterfaces/GeneralPlot.py:186]: this block absolutely should not be removed. It took me several days to get this far, and would be a shame to start over from scratch when this is inevitable fixed

File: ravenframework/OutStreams/PlotInterfaces/GeneralPlot.py [hunk 9]
```diff
@@ -136,204 +95,157 @@ def __init__(self):
...
+  def _returnSplitIndex(self, var, pltIndex, ind):
+    """
+      Function to check and return index for DataSet selection in filling variable values for non HistorySet
+      @ In, var, string, variable name to grab
+      @ In, pltIndex, int, plot index
+      @ In, ind, variable index
+      @ Out, split, str, key for DataSet
+    """
+
+    Split = self.__splitVariableNames(var, (pltIndex, ind))
```
Comment (PaulTalbot-INL) [inline ravenframework/OutStreams/PlotInterfaces/GeneralPlot.py:179]: Split is not a legal RAVEN name

File: ravenframework/Samplers/EnsembleForward.py [hunk 10]
```diff
@@ -18,30 +18,29 @@
...
 from .Stratified            import Stratified
 from .FactorialDesign       import FactorialDesign
 from .ResponseSurfaceDesign import ResponseSurfaceDesign
 from .CustomSampler         import CustomSampler
 from .. import GridEntities
-#Internal Modules End--------------------------------------------------------------------------------
+# Internal Modules End------------------------------------------------------------------------------
 
-class EnsembleForward(ForwardSampler):
+class EnsembleForward(Sampler):
```
Comment (PaulTalbot-INL) [inline ravenframework/Samplers/EnsembleForward.py:43]: This change is something that should have been discussed with the development team. Even if it appears there's no mechanical reason for a subclass to exist, it's at least worth checking with the existing team to see if it should be removed.

**Review by PaulTalbot-INL — CHANGES_REQUESTED** [review]
Review Part 1: beginning through `ravenframework/Samplers`. I'll keep going but leave this first part as a starter.

I will note there are a tremendous amount of personal preference changes in this PR; I would avoid this level of change in the future, and only fix items that are specifically in violation of the style guide, as if everyone did this we would rewrite half the code with each PR oscillating between linter/dev preferences.

File: ravenframework/Simulation.py [hunk 11]
```diff
@@ -688,78 +709,106 @@ def __readRunInfo(self,xmlNode,runInfoSkip,xmlFilename):
...
+      elif role == 'Optimizer' and self.ranPreviously:
+        # if simulation was run previously, Optimizer needs to be flushed
+        flushOptimizer = self.getEntity(entity, name)
+        flushOptimizer.flushOptimizer()
+        # now add to stepInputDict
+        stepInputDict[role] = flushOptimizer
+      elif role == 'Sampler' and self.ranPreviously:
+        # if simulation was run previously, Sampler needs to be flushed
+        flushSampler = self.getEntity(entity, name)
+        flushSampler.flushSampler()
```
Comment (PaulTalbot-INL) [inline ravenframework/Simulation.py:801]: wouldn't a common "flush" method name allow us to reduce the code added here? Is there a reason to preserve specific flush method names?

**Review by PaulTalbot-INL — CHANGES_REQUESTED** [review]
Remaining comments; the second half was much shorter than the first.

Consistently adding a newline after each return is a style choice that may or may not be universal; is there a pep8 guidance or something pushing for that? Where we don't generally have newlines within methods, this reduces viewability in my opinion.

File: ravenframework/MessageHandler.py (hunk 2 above)
Comment (dgarrett622) [reply ravenframework/MessageHandler.py:29]: This already existed in `utils`. Every reference I found to `NoMoreSamplesNeeded` used the one from `utils`.

File: ravenframework/MessageHandler.py [hunk 12]
```diff
@@ -352,5 +350,3 @@ def getHandler(name='default'):
   # NOTE: idk why, but h = _handlers.get(name, makeHandler(name)) does not work.
   # I think it's because it executes makeHandler(name) regardless of if name is present or not.
   return h
-
-builtins.getMessageHandler = getHandler
```
[the remark this answers is not in the document]
Comment (dgarrett622) [reply ravenframework/MessageHandler.py:356]: Not necessary for this PR. I don't like adding to the `builtins` which is why I made the change in the first place (`pylint` set up following the recommendations on the Wiki for VS Code doesn't like it as well). The `getMessageHandler` functionality is currently only used in 3 files: `MessageUser.py`, `mathUtils.py`, and `testMathUtils.py`. I'll revert it back.

File: ravenframework/MessageHandler.py (hunk 3 above)
Comment (dgarrett622) [reply ravenframework/MessageHandler.py:72]: No. I disagree with the previous implementation, but since it worked, I will put it back.

File: ravenframework/MessageHandler.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply ravenframework/MessageHandler.py:356]: Yeah, there's intentions to spread its use to more use cases, utils, and etc. I checked the logger package, and it looks like this is more or less what they do (or did) without telling you about it.

We need to try the logger package again at some point to see if there's massive slowdown from using it; I'd rather use the standard library than roll our own if we can get away with it; just need time to look at it.

File: ravenframework/MessageHandler.py (hunk 4 above)
Comment (PaulTalbot-INL) [reply ravenframework/MessageHandler.py:109]: Yeah, this is stuff that Andrea held over from Fortran; it was an ongoing debate on whether we should or not. I think we ended up going with pep8's extremely strong opinion on this one in the end.

File: ravenframework/MessageHandler.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply ravenframework/MessageHandler.py:356]: I am open to taking this change to the RAVEN group and going with whichever they want to do. I don't love putting stuff in the builtins either.

File: ravenframework/Databases/Database.py (hunk 1 above)
Comment (dgarrett622) [reply ravenframework/Databases/Database.py:139]: VS Code with `pylint` set up following the Wiki instructions doesn't like it. I tried to clean up as many of the `pylint` "problems" as I could and this is one of them.

File: ravenframework/DataObjects/DataSet.py [hunk 13]
```diff
@@ -2248,14 +2289,15 @@ def _generateHierPaths(self):
       # reconstruct path that leads to this ending
       path = [ending['prefix']]
       while ending['RAVEN_parentID'] != "None" and not pd.isnull(ending['RAVEN_parentID']):
-        _,ending = self.realization(matchDict={'prefix':ending['RAVEN_parentID']})
+        _,ending = self.realization(matchDict={'prefix': ending['RAVEN_parentID']})
```
[the remark this answers is not in the document]
Comment (dgarrett622) [reply ravenframework/DataObjects/DataSet.py:2292]: Fixed.

File: ravenframework/Models/HybridModels/HybridModel.py (hunk 5 above)
Comment (dgarrett622) [reply ravenframework/Models/HybridModels/HybridModel.py:186]: I created a new `resetHybridModel` method that encapsulates this.

File: ravenframework/Optimizers/GradientDescent.py (hunk 6 above)
Comment (dgarrett622) [reply ravenframework/Optimizers/GradientDescent.py:131]: Okay, I'll remove it.

File: ravenframework/Optimizers/GradientDescent.py (hunk 7 above)
Comment (dgarrett622) [reply ravenframework/Optimizers/GradientDescent.py:694]: Blame that on my PhD advisor, he was very particular about everything. I can put them back in if you really want them.

File: ravenframework/OutStreams/PlotInterfaces/GeneralPlot.py (hunk 8 above)
Comment (dgarrett622) [reply ravenframework/OutStreams/PlotInterfaces/GeneralPlot.py:186]: Okay, I'll put it back.

File: ravenframework/OutStreams/PlotInterfaces/GeneralPlot.py (hunk 9 above)
Comment (dgarrett622) [reply ravenframework/OutStreams/PlotInterfaces/GeneralPlot.py:179]: Renamed to firstSplit.

File: ravenframework/Samplers/EnsembleForward.py (hunk 10 above)
Comment (dgarrett622) [reply ravenframework/Samplers/EnsembleForward.py:43]: EAFP. I don't like keeping placeholders like this, but if you want it back I can put it back.

File: ravenframework/Simulation.py [hunk 14]
```diff
@@ -380,164 +366,195 @@ def XMLpreprocess(self,node,cwd):
...
-        if child.tag not in ['TestInfo']:
-          self.raiseAnError(IOError,'<'+child.tag+'> is not among the known simulation components '+repr(child))
-    # If requested, duplicate input
-    # ###NOTE: All substitutions to the XML input tree should be done BEFORE this point!!
-    if self.runInfoDict.get('printInput',False):
-      fileName = os.path.join(self.runInfoDict['WorkingDir'],self.runInfoDict['printInput'])
-      self.raiseAMessage('Writing duplicate input file:',fileName)
-      outFile = open(fileName,'w')
-      outFile.writelines(utils.toString(TreeStructure.tostring(xmlNode))+'\n') #\n for no-end-of-line issue
-      outFile.close()
```
[the remark this answers is not in the document]
Comment (dgarrett622) [reply ravenframework/Simulation.py:471]: Not sure what happened here. I'll put it back.

File: ravenframework/Simulation.py (hunk 11 above)
Comment (dgarrett622) [reply ravenframework/Simulation.py:801]: Hm, I'll have to look into this. I'm not sure how that would work with the relationship between Optimizers and Samplers.

File: ravenframework/Optimizers/GradientDescent.py (hunk 7 above)
Comment (PaulTalbot-INL) [reply ravenframework/Optimizers/GradientDescent.py:694]: Not worth reverting, just a note that may save you time next time!

File: ravenframework/Samplers/EnsembleForward.py (hunk 10 above)
Comment (PaulTalbot-INL) [reply ravenframework/Samplers/EnsembleForward.py:43]: Probably not worth reverting.

File: ravenframework/Simulation.py [hunk 15]
```diff
@@ -771,37 +820,41 @@ def initiateStep(self, stepName):
...
     stepInstance.takeAstep(stepInputDict)
-    #---------------here what is going on? Please add comments-----------------
+    # FilePrint has a finalize method that should be run at the end of the step,
+    # run this method for each instance
     for output in stepInputDict['Output']:
-      if "finalize" in dir(output):
-        output.finalize()
-    self.raiseAMessage('-'*2+' End step {0:50} '.format(stepInstance.name+' of type: '+stepInstance.type)+2*'-'+'\n')#,color='green')
+      if "_printer" in output.__dict__:
+        output._printer.finalize()
```
[the remark this answers is not in the document]
Comment (dgarrett622) [reply ravenframework/Simulation.py:833]: The original code was not finding the "finalize" method because it is inside the _printer object, so it was not performing as expected. The only "finalize" method I found in the code is in `FilePrint.py`. Do we anticipate other outputs with a "finalize" method? If so, I could create a "finalize" method in `PrintEntity.py` that calls the "finalize" method from `FilePrint.py` and go back to the original structure. This would remove the check on the protected object.

File: ravenframework/Simulation.py [hunk 16]
```diff
@@ -771,37 +820,41 @@ def initiateStep(self, stepName):
...
+    self.ranPreviously = True
 
   def stepSequence(self):
     """
       Return step sequence
       @ In, None
       @ Out, stepSequence, list(str), list of step in sequence
     """
+
     return self.__stepSequenceList
```
[the remark this answers is not in the document]
Comment (dgarrett622) [reply ravenframework/Simulation.py:858]: I like all of my return statements to be offset with whitespace. I find it more readable. Since I'm making edits to `Simulation.py` I'll get rid of it.

File: ravenframework/Simulation.py [hunk 17]
```diff
@@ -842,26 +927,24 @@ def generateAllAssemblers(self, objectInstance):
       neededobjs    = {}
       neededObjects = objectInstance.whatDoINeed()
       for mainClassStr in neededObjects.keys():
-        if mainClassStr not in self.entities.keys() and mainClassStr != 'internal':
-          self.raiseAnError(IOError,'Main Class '+mainClassStr+' needed by '+stp.name + ' unknown!')
+        if mainClassStr not in self.entities and mainClassStr != 'internal':
+          self.raiseAnError(IOError,'Main Class '+mainClassStr+' unknown!')
```
[the remark this answers is not in the document]
Comment (dgarrett622) [reply ravenframework/Simulation.py:931]: The variable "stp" is undefined in this method. Is it the same as "objectInstance"? If so, I can put it back in.

File: ravenframework/Simulation.py [hunk 18]
```diff
@@ -810,10 +863,41 @@ def run(self):
       @ In, None
       @ Out, None
     """
-    #to do list
-    #can we remove the check on the existence of the file, it might make more sense just to check in case they are input and before the step they are used
+    if self.jobHandler.completed:
+      # this must be False in order to set up the queue
+      self.jobHandler.completed = False
```
[the remark this answers is not in the document]
Comment (dgarrett622) [reply ravenframework/Simulation.py:868]: Hm, I would prefer to keep such a "reset" method closer to where the magic happens. Maybe I create a "reset" method in Simulation and then call it from PythonRaven if needed?

File: ravenframework/Steps/MultiRun.py [hunk 19]
```diff
@@ -44,94 +44,93 @@ def __init__(self):
       @ Out, None
     """
     super().__init__()
-    self._samplerInitDict = {} #this is a dictionary that gets sent as key-worded list to the initialization of the sampler
-    self.counter          = 0  #just an handy counter of the runs already performed
+    self._samplerInitDict = {} # dictionary that gets sent to the initialization of the sampler
+    self.counter          = 0  # counter of the runs already performed
+    self._outputCollectionLambda = None # lambda function list to collect the output without checking the type
```
[the remark this answers is not in the document]
Comment (dgarrett622) [reply ravenframework/Steps/MultiRun.py:49]: Done. Much nicer than aligning "=".

File: tests/framework/unit_tests/CustomDrivers/DemoPythonRaven.py [hunk 20]
```diff
@@ -49,8 +50,31 @@
 # uncomment for live plotting
 # plt.show()
 
-# run it again - this fails if not re-initialized!
-# raven.runWorkflow()
+# run workflow again to replicate results
+# change csv file name to avoid overwriting the same file
+printer = raven.getEntity('OutStreams', 'to_file')
+printer._printer.name = 'to_file2'
```
[the remark this answers is not in the document]
Comment (dgarrett622) [reply tests/framework/unit_tests/CustomDrivers/DemoPythonRaven.py:56]: Yep, I agree. I think we need another pull request to implement making changes like this. I hacked this together to create a test showing that you get the same results.

File: ravenframework/Simulation.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply ravenframework/Simulation.py:833]: oh, interesting. I think I agree with your assessment, the Entities should have the finalize method, and in those methods they should call finalize on their Interfaces. Given that this is only implemented for one kind of Entity right now, maybe it's not worth making a system out of it.

File: ravenframework/Simulation.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply ravenframework/Simulation.py:868]: I like that idea!

File: ravenframework/Simulation.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply ravenframework/Simulation.py:931]: If the name isn't in scope anymore, we can leave it off. A user should be able to figure out what step they're in based on other screen outputs ... maybe?

File: ravenframework/Simulation.py (hunk 11 above)
Comment (dgarrett622) [reply ravenframework/Simulation.py:801]: Okay, looks like a common "flush" method will work.

**Comment by PaulTalbot-INL** [comment]
Closes #1682

**Comment by PaulTalbot-INL** [comment]
An email to the user list is premature, but an email to the developer list noting this significant development is a good idea.
